app/src/lib/stores/helpers/infer-comparison-branch.ts

```diff
@@ -0,0 +1,129 @@
+  if (currentPullRequest !== null) {
+    return _getFeatureBranchOfPullRequest(branches, currentPullRequest)
+  }
+
+  const ghRepo = repository.gitHubRepository
+  if (ghRepo !== null) {
+    return ghRepo.fork && currentBranch !== null
+      ? _getDefaultBranchOfFork(branches, repository, ghRepo, currentBranch)
+      : _getDeafultBranchOfGithubRepo(branches, ghRepo)
+  }
+
+  return _getMasterBranch(branches)
+}
+
+/**
+ * Do **not** call this method; use `inferComparisonBranch` instead
+ *
+ * Returns `master`, the default branch of a Git repository
+ *
+ * @param branches The list of all branches for the repository
+ */
+export function _getMasterBranch(
+  branches: ReadonlyArray<Branch>
+): Branch | null {
+  return branches.find(b => b.name === 'master') || null
+}
+
+/**
+ * Do **not** call this method; use `inferComparisonBranch` instead
+ *
+ * Returns the default branch of the GitHub repository
+ *
+ * @param branches The list of all branches for the repository
+ * @param ghRepository The repository the branch belongs to
+ */
+export function _getDeafultBranchOfGithubRepo(
+  branches: ReadonlyArray<Branch>,
+  ghRepository: GitHubRepository
+): Branch | null {
+  return branches.find(b => b.name === ghRepository.defaultBranch) || null
+}
+
+/**
+ * Do **not** call this method; use `inferComparisonBranch` instead
+ *
+ * Returns the base branch of the given pull request.
+ *
+ * @param branches The list of all branches for the repository
+ * @param pr The pull request to use for finding the branch
+ */
+export function _getFeatureBranchOfPullRequest(
+  branches: ReadonlyArray<Branch>,
+  pr: PullRequest
+): Branch | null {
+  return branches.find(b => b.name === pr.base.ref) || null
+}
+
+/**
+ * Do **not** call this method; use `inferComparisonBranch` instead
+ *
+ * Returns the default branch of the fork if it's ahead of `currentBranch`.
+ * Otherwise, the default branch of the parent is returned.
+ *
+ * @param branches The list of all branches for the repository
+ * @param repository The repository the branch belongs to
+ * @param ghRepository
+ * @param currentBranch The branch we want the parent of
+ * @param getAheadBehind Callback function used to compute ahead/behind
+ */
+export async function _getDefaultBranchOfFork(
+  branches: ReadonlyArray<Branch>,
+  repository: Repository,
+  ghRepository: GitHubRepository,
+  currentBranch: Branch,
+  mockingAheadBehind?: (range: string) => IAheadBehind
+): Promise<Branch | null> {
+  const defaultBranchName = ghRepository.defaultBranch
+
+  if (defaultBranchName === null) {
+    // Fall back to master
+    return _getMasterBranch(branches)
+  }
+
+  const defaultBranch = _getDeafultBranchOfGithubRepo(branches, ghRepository)
+
+  if (defaultBranch === null) {
+    // Fall back to master
+    return _getMasterBranch(branches)
+  }
+
+  const range = revRange(currentBranch.tip.sha, defaultBranch.tip.sha)
+
+  // use the mock if it's passed - for testing only
```

Comment:
i'm not sure how i feel about passing in a param just to make tests pass. in cases like this, i've seen people do one of the following:
1. use `sinon` to get mocking functionality
1. pass the `getAheadBehind` function into the function directly so that it's only imported to be used in the public method and it's not directly exposed to the "private" method. this also makes it testable because you can mock what it returns in the tests
1. fully build the test object so that you can call `revRange` from it and call `getAheadBehind` on it
1. call getAheadBehind in the overseeing public method and pass in the result into this method

Reply:
I did something similar to item 2. I am still passing the `getAheadBehind` to the private function because I wanted to keep that logic as close to the place it's being used as possible. Hopefully it's a decent compromise.